fastseq/optimizer/fairseq/el_attention_optimizer.py

```diff
@@ -0,0 +1,1459 @@
+            if not self.training or (dropout_probability
+                    > self.decoder_layerdrop):
+                x, layer_attn = layer(
+                    x,
+                    encoder_state,
+                    encoder_out_v,
+                    (encoder_out.encoder_padding_mask
+                    if encoder_out is not None else None),
+                    incremental_state,
+                    self_attn_mask=self_attn_mask,
+                    self_attn_padding_mask=self_attn_padding_mask,
+                    need_attn=(idx == alignment_layer),
+                    need_head_weights=(idx == alignment_layer),
+                )
+                inner_states.append(x)
+                if layer_attn is not None and idx == alignment_layer:
+                    attn = layer_attn.float()
+
+        if attn is not None:
+            if alignment_heads is not None:
+                attn = attn[:alignment_heads]
+
+            # average probabilities over heads
+            attn = attn.mean(dim=0)
+
+        if self.layer_norm:
+            x = self.layer_norm(x)
+
+        # T x B x C -> B x T x C
+        x = x.transpose(0, 1)
+
+        if self.project_out_dim is not None:
+            x = self.project_out_dim(x)
+
+        return x, {'attn': attn, 'inner_states': inner_states}
+
+
+@replace(FairseqEncoderDecoderModel, use_el_attn)
+class FairseqEncoderDecoderModelV2(FairseqEncoderDecoderModel):
+    """class for encoder-decoder models.
+    Args:
+        encoder (FairseqEncoder): the encoder
+        decoder (FairseqDecoder): the decoder
+    """
+
+    def transpose_enc_dec_kv_proj (self):
+        for i in range (0, len(self.decoder.layers)):
+            self.num_heads = self.decoder.layers[i].encoder_attn.num_heads
+            self.head_dim = self.decoder.layers[i].encoder_attn.head_dim
+
+
+            self.decoder.layers[i].encoder_attn.k_proj_weight_t = (
+                    self.decoder.layers[i].encoder_attn.k_proj.weight
+                    .view(self.num_heads,
+                        self.head_dim, self.num_heads * self.head_dim)
+                    ).cuda()
+            self.decoder.layers[i].encoder_attn.k_proj_bias_t = (
+                    self.decoder.layers[i].encoder_attn.k_proj.bias
+                    .view(self.num_heads, self.head_dim, 1)
+                    ).cuda()
+
+            self.decoder.layers[i].encoder_attn.v_proj_weight_t = (
+                self.decoder.layers[i].encoder_attn.v_proj.weight
+                .view(self.num_heads, self.head_dim,
+                    self.num_heads * self.head_dim)
+                .transpose(1, 2)
+                .contiguous()
+                ).cuda()
+            self.decoder.layers[i].encoder_attn.v_proj_bias_t = (
+                self.decoder.layers[i].encoder_attn.v_proj.bias
+                .view(1, 1, self.num_heads * self.head_dim)
+                ).cuda()
+
+@replace(TransformerModel, use_el_attn)
```

Comment:
Line 502 - 512 can be removed?

Reply:
Looks we cancel another modification here, so have to keep it. At least line 509 to 512 are doing nothing.

Reply:
removed done